src/Client/Updater.php

```diff
@@ -644,6 +640,70 @@ public function download(string $target, ...$extra): PromiseInterface
             });
     }
 
+    /**
+     * Gets a target metadata object that contains the specified target.
+     *
+     * @param string $target
+     *   The path of the target file. Needs to be known to the most recent
+     *   targets metadata downloaded in ::refresh().
+     * @param \Tuf\Metadata\TargetsMetadata|null $targetsMetadata
+     *   The targets metadata to search or null. If null then the search will
+     *   start at the top level 'targets.json' file.
+     * @param string[] $searchedRoles
+     *   The roles that have already been searched.
+     *
+     * @return \Tuf\Metadata\TargetsMetadata|null
+     *   The target metadata with a match for the target, or null no match is
+     *   found.
+     */
+    protected function getMetadataForTarget(string $target, ?TargetsMetadata $targetsMetadata = null, array $searchedRoles = []): ?TargetsMetadata
+    {
+        if ($targetsMetadata === null) {
+            // If no target metadata is provided then start searching with the top level targets.json file.
+            /** @var \Tuf\Metadata\TargetsMetadata $targetsMetadata */
+            $targetsMetadata = TargetsMetadata::createFromJson($this->durableStorage['targets.json']);
+            if ($targetsMetadata->hasTarget($target)) {
```

Comment:
Shouldn't this `if` block come _after_ the check if `$targetsMetadata === null`, so that we return as quickly as possible?

Reply:
Only the top level `targets.json` needs to be searched here. In other calls we have already called `$newTargetsData->hasTarget($target)` when are looping through delegated and before we call this function recursively.  

If waited to do it here and we only had `$targetsMetadata` we won't have the information to know we should loop through the delegated roles in `$targetsMetadata` based on whether it is `terminating`.  `$targetsMetadata` can't tell us if this is for a terminating role only the metadata file that delegated to this role can tell us. 

You can set up the delegation incorrectly so that you make a terminating delegation but then delegate from that terminating delegation. We need to make sure we don't search those.

I have added a todo to #142 to test for the case.